### PR TITLE
fix(pp): route Qwen3.5 MoE pre-embedded inputs

### PR DESCRIPTION
**Pull request by HuiyingLi**
Summary:
- Route Qwen3.5 MoE pre-embedded tensors across pipeline stages.
- Keep multimodal pre-embedding on the separately sharded embedding module, avoiding duplicate text FSDP-root forwards.
- Add focused Qwen3.5 MoE PP routing and text-only state-adapter coverage.

Validation:
- 4x H100: PP2 x DP2, FSDP2, real multimodal inputs, 8 training steps; finite losses and non-zero finite embedding gradients.
- /opt/venv/bin/python -m pytest -q tests/unit_tests/models/qwen3_5_moe/test_qwen3_5_moe_model.py (53 passed)
- pytest -q tests/unit_tests/models/qwen3_5_moe (102 passed, 55 skipped; before removal of the isolated _pre_embed_only unit test)
- ruff check/format on changed Python files

### CODE REVIEW

File: nemo_automodel/components/models/qwen3_5_moe/model.py [hunk 1]
```diff
@@ -738,6 +751,227 @@ def init_weights(self, buffer_device: torch.device | None = None) -> None:
...
+        past_key_values: Any | None = None,
+        inputs_embeds: torch.FloatTensor | None = None,
+        labels: torch.LongTensor | None = None,
+        use_cache: bool | None = None,
+        logits_to_keep: int | torch.Tensor = 0,
+        output_hidden_states: bool | None = None,
+        cache_position: torch.Tensor | None = None,
+        padding_mask: torch.Tensor | None = None,
+        **kwargs: Any,
+    ) -> Qwen3_5MoeCausalLMOutputWithPast:
```
[the remark this answers is not in the document]
Comment (HuiyingLi) [reply nemo_automodel/components/models/qwen3_5_moe/model.py:876]: Addressed in 7a664ac9: documented the overloaded pipeline-stage input contract and tensor layouts, and widened input_ids to torch.Tensor to reflect token IDs or hidden states.

File: nemo_automodel/components/models/qwen3_5_moe/state_dict_adapter.py [hunk 2]
```diff
@@ -312,7 +314,7 @@ def store_native_key(native_key: str, tensor: Any) -> None:
                 continue
 
             match = re.match(
-                r"(?:model\.)?language_model\.layers\.(\d+)\.mlp\.experts\.(gate_up_proj|down_proj)$",
+                r"(?:model\.)?(?:language_model\.)?layers\.(\d+)\.mlp\.experts\.(gate_up_proj|down_proj)$",
```
[the remark this answers is not in the document]
Comment (HuiyingLi) [reply nemo_automodel/components/models/qwen3_5_moe/state_dict_adapter.py:317]: Addressed in 7a664ac9: added a text_only=True HF → native → HF expert round-trip covering model.layers keys, transposed shapes, and values.

**Comment by HuiyingLi** [comment]
Follow-up on the multimodal pre-embedding path (`edc7e3f1`): I removed the private `_pre_embed_only` dispatch and restored the direct `embed_tokens(input_ids)` call.

`apply_fsdp` already wraps `embed_tokens` as its own FSDP unit before wrapping the text root. In a 4xH100 PP2 x DP2 FSDP2 A/B run with real image inputs and 8 training steps, the direct path completed with finite losses and finite/non-zero embedding gradients. The private path also passed, but invoked the text root twice per microbatch (`pre_embed=16`, `decode=16`) versus one decode call on the direct path (`pre_embed=0`, `decode=16`). The direct path therefore keeps the same correctness while avoiding extra root hooks/collectives and hidden forward control flow.

**Comment by HuiyingLi** [comment]
Added CPU-safe text-only causal-LM execution coverage in `903a41c3` after `codecov/patch` reported 65.51% against an 80% target. The test uses a tiny `Qwen3_5MoeTextConfig` and executes `from_config`, weight initialization, a real forward pass, and backward with finite/non-zero embedding-gradient assertions. Local container result: `2 passed` together with the tie-policy guard.

**Review by akoumpa — APPROVED** [review]
